### DISCUSSION TITLE
Starting MariaDB at Boot

### DISCUSSION THREAD

**Discussion by l008com** [Homebrew/discussions#1741] [Casks] [answered]
I have a Mac mini running macOS 10.15 (Catalina) and I have installed mariadb 10.5.9 via homebrew.

Everything runs well and properly, with one exception: I cannot get the database to start up on system boot. It tries with the brew-installed launchd item. But it fails for some reason. Once logged in to the computer, if I run that very same command as root, the database starts right up and runs no problem.

I've had similar problems in the past that were caused by the permissions on `/usr/local/var/mysql` but that does not appear to be the problem here.

So here's the specifics. This is the launchd item I'm using to start the server. It *is* running, I can see the database try to start up in its error log.

```
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">

<plist version="1.0">
<dict>
	<key>Label</key>
		<string>homebrew.mxcl.mariadb</string>
	<key>ProgramArguments</key>
		<array>
			<string>/usr/local/opt/mariadb/bin/mysqld_safe</string>
			<string>--datadir=/usr/local/var/mysql</string>
		</array>
	<key>RunAtLoad</key>
		<true/>
	<key>WorkingDirectory</key>
		<string>/usr/local/var</string>
</dict>
</plist>
```

So that tries to start up the database, but fails. But when I run that very same command in the terminal after boot, the database starts right up fine. But if I forget to do this, of course the world will explode or at least something bad will happen.

`sudo /usr/local/opt/mariadb/bin/mysqld_safe --datadir=/usr/local/var/mysql
`
When the launchd plist tries to start up the server, the following is created in the server's error log:

```
210626 04:12:38 mysqld_safe Starting mariadbd daemon with databases from /usr/local/var/mysql
2021-06-26  4:12:39 0 [Note] /usr/local/opt/mariadb/bin/mariadbd (mysqld 10.5.9-MariaDB) starting as process 468 ...
2021-06-26  4:12:39 0 [Warning] Could not increase number of max_open_files to more than 256 (request: 32186)
2021-06-26  4:12:39 0 [Warning] Changed limits: max_open_files: 256  max_connections: 151 (was 151)  table_cache: 200 (was 2000)
2021-06-26  4:12:39 0 [Warning] Setting lower_case_table_names=2 because file system for /usr/local/var/mysql/ is case insensitive
/usr/local/opt/mariadb/bin/mariadbd: Please consult the Knowledge Base to find out how to run mysqld as root!
2021-06-26  4:12:39 0 [ERROR] Aborting
210626 04:12:39 mysqld_safe mysqld from pid file /usr/local/var/mysql/Home-Server.local.pid ended
```

There are a few `warnings` but the only error is just a message `Aborting`. It doesn't say *WHY* it is Aborting. Is one of the Warnings causing a problem? If so, which one? And why does the server start up fine if I do it from the terminal? Surely the max connections count and the case sensitivity of the filesystem are the same.

I am at a loss.

**Comment by SMillerDev** [comment]
Why not use `brew services`?

**Reply by l008com** [reply]
That launchd is from brew services. That's what it installs.

**Reply by SMillerDev** [reply]
And when you see it running, what does `brew services list` report?

**Reply by l008com** [reply]
"mariadb unknown root /Library/LaunchDaemons/homebrew.mxcl.mariadb.plist"

**Reply by SMillerDev** [reply]
And if you run it as sudo?

**Reply by l008com** [reply]
"mariadb started root /path..."

But it's only started because I manually started it, because it won't start on it's own.

**Reply by SMillerDev** [reply]
And how did you start it?

**Reply by l008com** [reply]
With the manual command in the original post above. Because it seems to die right away when started through launchd.

**Reply by SMillerDev** [reply]
But `brew services` thinks the job is started. What happens when you stop the manual command and do `sudo brew services restart mariadb`? That should just work as you're used to or at least post an error message why it fails.

**Reply by l008com** [reply]
I'm not on the server machine right now. I'll give that a try tomorrow and get back to you. But the error log from the launchd startup command is also up in the original post if that is useful at all.

**Reply by l008com** [reply]
Ok on a fresh restart, `sudo brew services list` gives me:
`mariadb started root /Library/LaunchDaemons/homebrew.mxcl.mariadb.plist`

Even though it is not running. I still have to manually start mariadb to get it running. 

When I run 1sudo brew services restart mariadb`, it thinks it stops it, then it thinks it starts it back up. But it's still not running until I manually do:

`sudo /usr/local/opt/mariadb/bin/mysqld_safe --datadir=/usr/local/var/mysql`

**Reply by l008com** [reply]
Oh and I was just in the system.log for something else and discovered that the launchd plist not only tries to start mariadb at boot, but it keeps trying to start it every 10 seconds continuously, even after I manually start it. Also I don't have "KeepAlive" enabled in this .plist so basically none of this makes any sense at all. 

... so I just looked at the .plist and it does in fact have KeepAlive enabled. I think this is new since I restarted with `brew services`, I think it replaced the launchd .plist file. In other words I think I removed the option and what I tried above restored to a default state, a default state that still doesn't actually work.

**Comment by l008com** [comment, marked as answer]
I found a solution! I'm not sure how or why it works, but if you add 'sudo' to the command launchd runs, mariadb is able to start up normally! This doesn't make sense because these launchdeamons are already being run as root, so adding sudo shouldn't DO anything. Yet, it fixes it! Go figure.

```
...
<key>ProgramArguments</key>
	<array>
		<string>/usr/bin/sudo</string>
		<string>/usr/local/opt/mariadb/bin/mariadbd-safe</string>
		<string>--datadir=/usr/local/var/mysql</string>
	</array>
...
```
